centaur/src/main/scala/centaur/test/Test.scala

```diff
@@ -419,6 +425,104 @@ object Operations extends StrictLogging {
+      val ioActualOutputs: IO[Map[String, JsValue]] = CentaurCromwellClient.outputs(submittedWorkflow) map { _.asFlat.stringifyValues }
+
+      ioActualOutputs flatMap { actualOutputs =>
+        val expected = expectedOutputs.toSet
+        val actual = actualOutputs.toSet
+
+        lazy val inActualButNotInExpected = actual.diff(expected)
+        lazy val inExpectedButNotInActual = expected.diff(actual)
+
+        if (!workflow.allowOtherOutputs && inActualButNotInExpected.nonEmpty) {
+          val message = s"In actual outputs but not in expected and other outputs not allowed: ${inActualButNotInExpected.mkString(", ")}"
+          IO.raiseError(CentaurTestException(message, workflow, submittedWorkflow))
+        } else if (inExpectedButNotInActual.nonEmpty) {
+          val message = s"In expected outputs but not in actual: ${inExpectedButNotInActual.mkString(", ")}"
+          IO.raiseError(CentaurTestException(message, workflow, submittedWorkflow))
+        } else {
+          IO.unit
+        }
+      }
+    }
+  }
+
+  def validateLabels(submittedWorkflow: SubmittedWorkflow,
+                     workflow: Workflow,
+                     workflowRoot: String): Test[Unit] = new Test[Unit] {
+    override def run: IO[Unit] = {
+      import centaur.test.metadata.WorkflowFlatLabels._
+
+      val workflowIdLabel = ("cromwell-workflow-id", JsString(s"cromwell-${submittedWorkflow.id}"))
+      val expectedLabels: List[(String, JsValue)] = workflowIdLabel ::
+        selectAndProcessExpectationFlatMetadata(workflow, "labels.", submittedWorkflow.id, workflowRoot)
+      val ioActualLabels: IO[Map[String, JsValue]] = CentaurCromwellClient.labels(submittedWorkflow) map { _.asFlat.stringifyValues }
+
+      ioActualLabels flatMap { actualLabels =>
+        val diff = expectedLabels.toSet.diff(actualLabels.toSet)
+        if (diff.nonEmpty) {
+          val message = s"In expected labels but not in actual: ${diff.mkString(", ")}"
+          IO.raiseError(CentaurTestException(message, workflow, submittedWorkflow))
+        } else {
+          IO.unit
+        }
+      }
+    }
+  }
+
+  /** Compares logs filtered from the raw `metadata` endpoint with the `logs` endpoint. */
+  def validateLogs(submittedWorkflow: SubmittedWorkflow, workflow: Workflow): Test[Unit] = new Test[Unit] {
+    val suffixes = Set("stdout", "shardIndex", "stderr", "attempt", "backendLogs.log")
+
+    def removeSubworkflowKeys(flattened: Map[String, JsValue]): Map[String, JsValue] = {
+      val subWorkflowIdPrefixes = flattened.keys.filter(_.endsWith(".subWorkflowId")).map(s => s.substring(0, s.lastIndexOf('.')))
+      flattened filter { case (k, _) => !subWorkflowIdPrefixes.exists(k.startsWith) }
+    }
+
+    def filterMetadata(flattened: Map[String, JsValue]): Map[String, JsValue] = removeSubworkflowKeys(flattened).filter {
+      case (k, _) => k == "id" || suffixes.exists(s => k.endsWith("." + s) && !k.contains(".outputs.") && !k.startsWith("outputs."))
```

Comment:
except for the leading dot in `".outputs."` which is kinda important

Reply:
Am I doing this wrong? I think they're equivalent:

| input | `!k.contains(".outputs.") && !k.startsWith("outputs.")` | `!k.contains("outputs.")` |
|--------------|-------------------------------------------------------|-------------------------|
| `".outputs."`  | false                                                 | false                   |
| `"outputs."`   | false                                                 | false                   |
| `".outputs"`   | true                                                 | true                   |
| `"outputs"`   | true                                                 | true                   |

Reply:
I guess it's better because of possible things like this:

| input | `!k.contains(".outputs.") && !k.startsWith("outputs.")` | `!k.contains("outputs.")` |
|--------------|-------------------------------------------------------|-------------------------|
| `"blahoutputs."`  | false                                                 | true                   |